### ISSUE TITLE
TempDatabase ignores $name in some circumstances

### ISSUE THREAD

**Issue by MasonD** [silverstripe/silverstripe-framework#8861]
## Affected Version

Silverstripe 4.3, probably all of 4

## Description

`TempDatabase->rebuildTables()` calls `->requireTable()` on every DataObject, but `DataObject->requireTable()` calls `DB::requireTable()`, which does not support named schemas, instead performing the manipulation on the default schema. Thus, creating a TempDatabase with any connection name besides default results in buggy behaviour, where builds are performed on the default connection instead.

## Steps to Reproduce

This is tricky. I encountered this problem when attempting to change my tests to use SQLite3Database instead of the rest of the site's configuration, by creating a second config with
```
DB::setConfig([
    "type" => 'SQLite3Database',
    "username" => DB::getConfig()['username'],
    "password" => DB::getConfig()['password'],
    "path" => ':memory:'
], 'sqlite');
```
If you then subclass `SapphireTest` and override `tempDB` like so:
```
public static function tempDB()
{
    if (!static::$tempDB) {
        static::$tempDB = TempDatabase::create('notdefault');
    }
    return static::$tempDB;
}
```
Add an empty test, run that test, and you will begin encountering full build output clogging up the test results. This is because the builds are called on the `default` connection, which hasn't had `->quiet()` called on it.

**Comment by ScopeyNZ** [comment]
Hmmmn. It looks like the build of the test DB is originally called on the property directly and doesn't use the getter. I wonder if you could overcome the problem by setting the `$tempDB` in an extension of `setupBeforeClass`:

```
public static function setUpBeforeClass()
{
    static::$tempDB = TemptDatabase::create('notdefault');
    parent::setUpBeforeClass();
}
```

**Comment by MasonD** [comment]
I don't believe that's what's happening here. `$tempDB` is null by default, and only ever set in `tempDB()`, so any calls on the property directly are either going to error or have the correct temp db. The issue here is that `TempDatabase::rebuildTables` calls `DataObject::requireTable()`, which calls `DB::require_table()`, which calls `requireTable` on the schema returned from `DB::get_schema()` with no arguments passed, which always defaults to 'default'.
